### DISCUSSION TITLE
Nutrition Information

### DISCUSSION THREAD

**Discussion by dfrankland** [cooklang/spec#52] [Ideas]
It'd be great to specify the nutrition of each ingredient and have that be tabulated to the total nutrition and per serving nutrition.

**Comment by edibotopic** [comment]
This would be great. If calculations from input macronutrient data (protein, lipid, carbs) are possible in the program then they could be converted to approximate caloric content using [Atwater factors](https://esha.com/blog/calorie-calculation-country/#:~:text=In%20the%20U.S.%2C%20most%20manufacturers,Atwater.).

**Reply by pubmania** [reply]
While I am unsure if I would want the cooklang itself to have the ability to Parse nutrition data, it can be achieved for smaller projects fairly easily as I have done on my small site using the nutritionix api.

They are quite generous on limits for free use. 

The advantage of this approach in my opinion is one does not have to manually enter the nutrition info on the recipe. Secondly, information from API is far more detailed than we could manually capture anyway, allowing for a better standardised visualisation of nutritional info as can be seen below:

![Screenshot_20250314_002215_Firefox.jpg](https://github.com/user-attachments/assets/f7cf7648-0abf-4ba0-93ef-2082ffbad17c)

**Reply by chrbala** [reply]
I'm currently working on building out my own API like this. I could potentially be interested in vending something to cooklang apps if there's interest.

**Reply by dfego** [reply]
https://support.cronometer.com/hc/en-us/articles/360018239472-Data-Sources

This app uses a few data sets. It looks like at least the USDA one is available to download and use.

**Reply by pubmania** [reply]
>  It looks like at least the USDA one is available to download and use.

The USDA data is available as API directly from official sources: https://fdc.nal.usda.gov/api-guide.html

**Comment by dubadub** [comment]
Yeah, definitely this was one of my biggest motivators to start Cooklang. It's paradoxical that while the nutrition balance is so important for the health, we have very little feedback on how we're doing.

Slowly we're getting there. But first we need to figure out [different units](https://github.com/cooklang/spec/issues/4) and [ingredient density](https://github.com/cooklang/spec/issues/23).

**Reply by azlekov** [reply]
Hey @dubadub unfortunately the links here

> Slowly we're getting there. But first we need to figure out [different units](https://github.com/cooklang/spec/issues/4) and [ingredient density](https://github.com/cooklang/spec/issues/23).

 are not working anymore. Is there any update on Nutrition Information?

I was thinking to use the metadata to put there all Nutrition Information, but looks like most of the parsers supports just string as a value, so I cannot group them. My idea was:

```
>> nutrition facts: [calories: 221kcal, carbohydrates: 52g, protein: 6g, fat: 1g, cholesterol: 2mg, sodium: 60mg, fiber: 4g, sugar: 41g]
```

Something which can be easily be parseable to dictionary.

The approach where in the meta I put all the facts in a separate item will require some additional work to group them for nice visualization later.


```
>> calories: 221kcal
>> carbohydrates: 52g
>> protein: 6g
>> fat: 1g
>> cholesterol: 2mg
>> sodium: 60mg
>> fiber: 4g
>> sugar: 41g
```

**Reply by dubadub** [reply]
> Hey @dubadub unfortunately the links here
> are not working anymore. Is there any update on Nutrition Information?

No updates yet, here is one of the links https://github.com/cooklang/spec/discussions/51 (changed after moving from issues to discussions), second one is gone.

In long-term I think there will be two approaches for nutrition facts:

1. Generating estimates based on ingredients data (with some optimism that we can make computers do their job)
2. Using canonical metadata like you mentioned. Canonical keys for this not defined yet. You can share them with me or update this PR https://github.com/cooklang/spec/pull/69/files#diff-683343bdf93f55ed3cada86151abb8051282e1936e58d4e0a04beca95dff6e51R142-R152 to include this into Roadmap. After that we can update CLI and app to display this data nicely and include it into main spec.

**Reply by Cabalist** [reply]
@dubadub  To your point on optimism this looks to be an interesting service on a small scale:
https://developer.edamam.com/edamam-nutrition-api

I don't think it is super relevant from a sustainable open source project standpoint but having a connector I could write that would pass it .cook files, get json back and then inserting it into the .cook files (in a canonical way) seems like something that just requires the 'canonical way' to be decided on.  :)


It returns something like this:
```python
{
    'yield': 6,
    'calories': 4453,
    'cautions': ['SULFITES', 'FODMAP'],
    'dietLabels': [],
    'totalWeight': 1839.536434375,
    'healthLabels': None,
    'totalNutrients': {
        'K': {
            'unit': 'mg',
            'label': 'Potassium, K',
            'quantity': 1578.6452949687496
        },
        'P': {
            'unit': 'mg',
            'label': 'Phosphorus, P',
            'quantity': 2115.28593625
        },
        'CA': {
            'unit': 'mg',
            'label': 'Calcium, Ca',
            'quantity': 3018.609738755267
        },
        'FE': {
            'unit': 'mg',
            'label': 'Iron, Fe',
            'quantity': 32.106330479852524
        },
        'MG': {'unit': 'mg', 'label': 'Magnesium, Mg', 'quantity': 233.821823},
        'NA': {
            'unit': 'mg',
            'label': 'Sodium, Na',
            'quantity': 6249.6787184375
        },
        'ZN': {
            'unit': 'mg',
            'label': 'Zinc, Zn',
            'quantity': 14.591718511249999
        },
        'FAT': {
            'unit': 'g',
            'label': 'Total lipid (fat)',
            'quantity': 248.98952886858143
        },
        'NIA': {'unit': 'mg', 'label': 'Niacin', 'quantity': 30.612364862375},
        'FAMS': {
            'unit': 'g',
            'label': 'Fatty acids, total monounsaturated',
            'quantity': 64.44136919228126
        },
        'FAPU': {
            'unit': 'g',
            'label': 'Fatty acids, total polyunsaturated',
            'quantity': 10.320202799843749
        },
        'RIBF': {
            'unit': 'mg',
            'label': 'Riboflavin',
            'quantity': 3.537692636937499
        },
        'THIA': {
            'unit': 'mg',
            'label': 'Thiamin',
            'quantity': 4.44658187165625
        },
        'VITC': {
            'unit': 'mg',
            'label': 'Vitamin C, total ascorbic acid',
            'quantity': 41.238846681249996
        },
        'VITD': {
            'unit': 'µg',
            'label': 'Vitamin D (D2 + D3)',
            'quantity': 1.5057465500000002
        },
        'CHOLE': {
            'unit': 'mg',
            'label': 'Cholesterol',
            'quantity': 301.26293072120785
        },
        'FASAT': {
            'unit': 'g',
            'label': 'Fatty acids, total saturated',
            'quantity': 72.11352240053547
        },
        'FATRN': {
            'unit': 'g',
            'label': 'Fatty acids, total trans',
            'quantity': 0
        },
        'FIBTG': {
            'unit': 'g',
            'label': 'Fiber, total dietary',
            'quantity': 23.323686719276683
        },
        'FOLAC': {
            'unit': 'µg',
            'label': 'Folic acid',
            'quantity': 654.8739841875
        },
        'FOLFD': {
            'unit': 'µg',
            'label': 'Folate, food',
            'quantity': 383.87059521875
        },
        'SUGAR': {
            'unit': 'g',
            'label': 'Sugars, total',
            'quantity': 27.4451861503125
        },
        'VITK1': {
            'unit': 'µg',
            'label': 'Vitamin K (phylloquinone)',
            'quantity': 99.60489525312502
        },
        'WATER': {'unit': 'g', 'label': 'Water', 'quantity': 831.965346323125},
        'CHOCDF': {
            'unit': 'g',
            'label': 'Carbohydrate, by difference',
            'quantity': 392.25124516902036
        },
        'FOLDFE': {
            'unit': 'µg',
            'label': 'Folate, DFE',
            'quantity': 1498.00685403125
        },
        'PROCNT': {
            'unit': 'g',
            'label': 'Protein',
            'quantity': 156.83993407281253
        },
        'TOCPHA': {
            'unit': 'mg',
            'label': 'Vitamin E (alpha-tocopherol)',
            'quantity': 11.041630610937501
        },
        'VITB12': {
            'unit': 'µg',
            'label': 'Vitamin B-12',
            'quantity': 7.3615287725
        },
        'VITB6A': {
            'unit': 'mg',
            'label': 'Vitamin B-6',
            'quantity': 0.7601271169375
        },
        'VITA_RAE': {
            'unit': 'µg',
            'label': 'Vitamin A, RAE',
            'quantity': 801.9469298125
        },
        'ENERC_KCAL': {
            'unit': 'kcal',
            'label': 'Energy',
            'quantity': 4453.754064853932
        },
        'SUGAR.added': {
            'unit': 'g',
            'label': 'Sugars, added',
            'quantity': 10.479000000000001
        }
    }
}
```

**Comment by dfego** [comment]
It's been quite some time since this was discussed. Are there any concrete steps you might suggest someone could take to help push this forward? I'm willing to spend some time trying to contribute!

**Reply by dubadub** [reply]
That's great! Actually, just in last newsletter I outlined my plan about that. Let me paste it in next message.

**Reply by dubadub** [reply]
## Cook Plan

I think there shouldn't be anything fancy—just support for linking other recipes from a recipe. That way, a cook plan is just another recipe (a meta-recipe) that contains links to other recipes and supports some formatting.  

Something like this:

<img width="848" alt="image" src="https://github.com/user-attachments/assets/cd4ec25f-a0a1-4955-8dda-2da8c25374d9" />

You probably see that this format is universal—it can also be used as a meal journal for tracking food intake. Plus, we can ask computers to generate these plans for us, for example based on metadata. Pretty cool, huh?

As first step, I drafted a proposal for linking recipes here: https://github.com/cooklang/spec/pull/118. A while ago, we had a discussion on the Discord server about this feature, and based on the use cases mentioned, I think the best approach is not to introduce new syntax but to detect that an ingredient is a link when it looks like a file path.

## Plain Text Database  

We need to store some data that we can later use for decision-making. Since I hate complexity, I want it to be as simple and portable as possible. What if we have a db folder with nested folders with data in YAML files? 

We can build a library that provides a simple key/value interface to these YAML files. Scripts could retrieve data from the database like this:  

<img width="585" alt="image" src="https://github.com/user-attachments/assets/b2b7fbbb-f753-4ef9-a026-572b085ba1f5" />

The part with slashes relates to paths, and the dots relate to the YAML structure.  
Of course, we could also have a shared community database, accessible both locally and remotely, for common data like nutrition values.  

Why reinvent the wheel? I checked various query languages, including a cool project like yq, which allows querying YAML files. But oh boy, it's too ugly and would make scripting unnecessarily convoluted.  

## Reports
It would be nice to connect these plans and data, right? So, we need reports.  

A report is essentially a scripted template that outputs markdown. The scripting context should have access to a parsed plan/recipe, the database, and some preloaded helpers. That way, we can calculate nutrition values, glycemic index, costs—basically anything—and present it as a simple document.  

We can't use Rust for this because that would require users to install a Rust compiler. But there are projects that allow compiling lightweight versions of interpreted languages like Ruby or Python from Rust itself (e.g., https://www.artichokeruby.org). Since I find Ruby adorable, that would probably be my first choice.  

A report could look something like this (ERB format):  

<img width="855" alt="image" src="https://github.com/user-attachments/assets/ee7cbf42-4206-4dc3-8047-36c6348a1275" />

We can create handy helpers and even generate inline image data.  

Since reports are just files, users can share them with each other and create custom ones for different diets, flagging issues, etc.

We should also have a built-in library for common tasks like value aggregation and unit conversion.

**Reply by dubadub** [reply]
I guess, it's not the most straightforward way to introduce nutrition values, but it will give us absolute flexibility in extracting information from recipes and then allow us to make smart meal plans with any possible constraints (diet, ingredient shelf-life, cost, etc).

**Reply by dubadub** [reply]
As of progress, I'm trying to get cook plans into the spec and implement parser support for it. You can start the plain text database part. What do you think?

**Reply by dfego** [reply]
Sure, I think I can wrap my head around the plaintext/YAML database part enough to take a stab at it. Once I dive into the code I can create a draft PR. Step 1 would be updating the parser (https://github.com/cooklang/cooklang-rs), right?

Related to reports, I'm not entirely sure I understand why Rust couldn't be used to parse templates. I understand why we wouldn't want anyone to have to install a compiler, but I don't understand why having templates that get parsed would require someone to have a compiler installed.

**Reply by dubadub** [reply]
Actually, I was thinking about having a separate library for that, you can start a repo and we will transfer it into the organisation. 

As for templates, I want to make them scriptable, so we would need to create a runtime for them. Or maybe I'm missing something, can share your thoughts?

**Reply by dfego** [reply]
Re: separate library, can do.

What do you mean by scriptable? I'm familiar with code that works on templates (e.g. like https://pkg.go.dev/text/template). Regardless of how complex the input is, theoretically the parser should be able to take a template and plaintext data as input and produce output without need to recompile. But maybe I'm missing something!

**Reply by dfego** [reply]
Separate question, so I don't fall into any pitfalls. You mentioned [yq](https://github.com/mikefarah/yq) as an example of being too complex. I've used `jq` before and agree the syntax is a bit complicated, though I always took at that as somewhat necessary due to the complexity of what it's doing.

How would you envision it being even simpler? You specifically mentioned it just having a key/value interface, so:

- No support for arrays
- No support for filters

If that's the path, I can forge ahead with some confidence.

**Reply by dubadub** [reply]
Yeah, I was thinking about key/value because I don't see any need to query collections of values from the database. In template where we use it we know the ingredient and know what kind of data we need. Hence jq syntax is just too much

**Reply by dubadub** [reply]
Well, the returned value can still be array or hashmap though

**Reply by dubadub** [reply]
> What do you mean by scriptable? I'm familiar with code that works on templates (e.g. like https://pkg.go.dev/text/template). Regardless of how complex the input is, theoretically the parser should be able to take a template and plaintext data as input and produce output without need to recompile. But maybe I'm missing something!

Looks interesting, definitely can be an option and make things easier. I'm wondering if rust supports something similar.

**Reply by dfego** [reply]
I'm currently working on this a bit and running into some interesting challenges since I'm making it generic. Apparently YAML is _very_ permissive in what it allows in, for example, key names. From what I can tell you can put basically any character imaginable in there. That creates the challenge of how to specify a string of keys (e.g. `market.cost.tesco`) in a string form with some kind of delimiter on our end, since `market.cost.tesco` could literally be the name of a key.

A more correct way to handle this is to have the interface explicitly take in a vector of relevant strings, i.e. `vec!["market", "cost", "tesco"]`. If I were to do that, you could enforce in the cooklang code what the format looks like and code to that.

tl;dr YAML permissive, creates ambiguity in the example use case. I'm currently moving towards making the actual API slightly more correct and generic, and leaving converting something `market.cost.tesco` into `vec!["market", "cost", "tesco"]` as an exercise for the user of the library, i.e. cooklang.

Just using this as a forum to dump some of my thoughts on the matter and also point out the direction I'm moving in case you have any issues!

**Reply by dubadub** [reply]
Interesting observation. Yes, good point, let's accept vectors. Are you on our Discord server? we may discuss things there and others can chip in?

**Comment by gvlx** [comment]
Nutritional aspects depend on the actual ingredients. Maybe a link with https://world.openfoodfacts.org/ on the client would be interesting?

And don't forget https://prices.openfoodfacts.org/ for the store/aisle info and purchasing price.

**Comment by romainhild** [comment]
I just discovered Cooklang and it looks awesome ! 
What are the latest news on the nutrition information ?

It seems to me that there are two different things discussed here, how to get the information and how to store them in the file. 
For the first point, it is not really a specification of the language, I guess that each app using the language could have a different approach for this (looking at a DB stored locally or using an API, or just looking at the file).
For the second point, it would be great if there was an official way to store directly the information in the file. It would be a starting point for the different apps to display those information. This would require the user to input the information himself but now, there are plenty of website that give them directly (I'm thinking of Hellofresh for example).

As proposed previously, I think the metadata would be the place, something like:
```
nutrition:
  calories: 776 kcal
  fat: 40.7 g
  saturated fat: 13 g
  carbohydrates: 75.4 g
  sugar: 11.1 g
  protein: 21.6 g
  fiber: 6.7 g
  sodium: 2.1 g
```
which is what is already done when parsing a HelloFresh recipe with cook.md.

**Reply by dubadub** [reply]
Yes, it's there in the import tool. Need to update parser to recognise this and scale with the recipe before showing this info in the apps.
